### PR TITLE
Implement LWG-3591 and LWG-3592 for `lazy_split_view`

### LINKED ISSUE #2401: LWG-3591 `lazy_split_view<input_view>::inner-iterator::base() &&` invalidates outer iterators
LWG-3591 `lazy_split_view<input_view>::`*`inner-iterator`*`::base() &&` invalidates outer iterators

This was voted into the Working Paper in October 2021, see #2236.

### LINKED ISSUE #2402: LWG-3592 `lazy_split_view` needs to check the simpleness of `Pattern`
LWG-3592 `lazy_split_view` needs to check the simpleness of `Pattern`

This was voted into the Working Paper in October 2021, see #2236.

### PR DESCRIPTION
**Pull request by cristeigabriel**
Fixes #2401 and fixes #2402.

### CODE REVIEW

**Comment by cristeigabriel** [comment]
I notice there's issue with the tests and I will be looking into it very soon.

File: stl/inc/ranges [hunk 1]
```diff
@@ -1947,7 +1947,7 @@ namespace ranges {
                 return _Current;
             }
             _NODISCARD constexpr iterator_t<_Base> base() && noexcept(
-                is_nothrow_move_constructible_v<iterator_t<_Base>>) /* strengthened */ {
+                is_nothrow_move_constructible_v<iterator_t<_Base>>) requires forward_range<_Vw> {
```
Comment (miscco) [inline stl/inc/ranges:1950]: So I was wonderung why this fails in the `transform_view` tests. You are in the wrong view here. 

This is the `transform_view::iterator` and not the `lazy_split_view::inner_iterator`. Please revert this change here.

You can find those functions as line 3660-3667

File: stl/inc/ranges [hunk 2]
```diff
@@ -4404,7 +4404,7 @@ namespace ranges {
             }
 
             _NODISCARD constexpr iterator_t<_Base> base() && noexcept(
-                is_nothrow_move_constructible_v<iterator_t<_Base>>) /* strengthened */ {
+                is_nothrow_move_constructible_v<iterator_t<_Base>>) requires forward_range<_Vw> {
```
Comment (miscco) [inline stl/inc/ranges:4407]: That one should be reverted. this is `elements_view` not `lazy_split_view`

**Review by miscco — COMMENTED** [review]
You will also need to adopt the tests in `STL\tests\std\tests\P0896R4_views_lazy_split`

Please have a look in the [Readme](https://github.com/microsoft/STL#how-to-run-the-tests-with-a-native-tools-command-prompt) on how to build the STL and then run those specific tests

**Comment by miscco** [comment]
Here are the links to the relevant code lines for this issue:

https://github.com/microsoft/STL/blob/2f9f567ffe1087353ecacc5a03f81c395a1922e1/stl/inc/ranges#L3673-L3675

It would be super awesome if you could also tackle https://github.com/microsoft/STL/issues/2402 while you are in the vicinity.

The relevant lines are 
https://github.com/microsoft/STL/blob/2f9f567ffe1087353ecacc5a03f81c395a1922e1/stl/inc/ranges#L3757-L3764

and https://github.com/microsoft/STL/blob/2f9f567ffe1087353ecacc5a03f81c395a1922e1/stl/inc/ranges#L3773-L3776

Note that we use `_Pat` for the type of the pattern so this should be `&& _Simple_view<_Pat>`

It is totally fine if you do not want to tackle both

**Comment by cristeigabriel** [comment]
Thank you very much! I'm on it (I'll be taking on what you asked me to), sorry for my lack of experience and thank you very much for being so helpful!

**Comment by miscco** [comment]
> Thank you very much! I'm on it (I'll be taking on what you asked me to), sorry for my lack of experience and thank you very much for being so helpful!

No worries, you can only get experience by trying it out and the stl is a complicated code base. If you need any help just ask right away

**Comment by cristeigabriel** [comment]
A peculiarity, using the following from the official LLVM GitHub releases page:
```
C:\Users\gabriel\Documents\GitHub\STL>clang-format --version
clang-format version 13.0.0
```

Here's what clang-format says:
```
C:\Users\gabriel\Documents\GitHub\STL>clang-format --dry-run -Werror stl/inc/ranges
stl/inc/ranges:2464:30: error: code should be clang-formatted [-Wclang-format-violations]
            return (_STD min) (_Length, static_cast<decltype(_Length)>(_Count));
                             ^
stl/inc/ranges:2469:30: error: code should be clang-formatted [-Wclang-format-violations]
            return (_STD min) (_Length, static_cast<decltype(_Length)>(_Count));
                             ^
stl/inc/ranges:2537:51: error: code should be clang-formatted [-Wclang-format-violations]
                    _Count            = (_STD min) (_RANGES distance(_Range), _Count);
                                                  ^
stl/inc/ranges:2795:48: error: code should be clang-formatted [-Wclang-format-violations]
                const auto _Offset = (_STD min) (_RANGES distance(_Range), _Count);
                                               ^
stl/inc/ranges:2816:44: error: code should be clang-formatted [-Wclang-format-violations]
            const auto _Offset = (_STD min) (_RANGES distance(_Range), _Count);
                                           ^
stl/inc/ranges:2907:40: error: code should be clang-formatted [-Wclang-format-violations]
                    _Count = (_STD min) (_RANGES distance(_Range), _Count);
                                       ^
```

Am I missing something? If so, sorry.

**Comment by miscco** [comment]
Thanks alot, formatting is still a pain point.

It also looks like the formatter is not too happy with some of the code but I am not sure whether it warrants turning it off.

I am slightly terrified that all Tests pass. I am going to bed now, but I hope to make some suggestions where to add Tests tomorrow if @CaseyCarter doesnt beat me to it

**Comment by fsb4000** [comment]
add `Fixes #2402` in your first post

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by CaseyCarter** [comment]
@cristeigabriel , @StephanTLavavej: I pushed a change to extract the boolean expression template argument out into a `constexpr bool` variable to make it clearer to clang-format that `_Outer_iter < /* stuff */ >` is a type and not an expression.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing these LWG issue resolutions and congratulations on your first microsoft/STL commit! :tada: :heart_eyes_cat: :white_check_mark:

This will ship in VS 2022 17.2 Preview 2.

**Pull request merged**